### PR TITLE
8344306: RISC-V: Add zicond

### PR DESCRIPTION
**Pull request by robehn**
Hi, please consider.

In cpu models we save ~1 cycle per removed branch.
This patch removes ~0.1% of branches in generic C2 generated code.
We should probably investigate if we can improve/add peephole optimization to remove more branches.

As the C1 cmov code is a bit tricky I left that as a followup.

I added gtests for the cmovs.
(we should add coverage for more of masm in this gtest suit)
Pro tip, invoke the gtestLauncher directly (you only need to build exploded):
`gtestLauncher -jdk build/linux-riscv64-server-fastdebug/jdk/ --gtest_break_on_failure --gtest_filter="*RiscV*" -XX:+UnlockDiagnosticVMOptions -XX:+UseZicond -XX:ParallelGCThreads=1 -XX:ConcGCThreads=1 -XX:CICompilerCount=2`

Tested on Spacemit X60, gtests and tier1.

Thanks, Robbin

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8344306](https://bugs.openjdk.org/browse/JDK-8344306): RISC-V: Add zicond (**Enhancement** - P4)

### Reviewers
 * [Hamlin Li](https://openjdk.org/census#mli) (@Hamlin-Li - **Reviewer**) Review applies to [3bae7423](https://git.openjdk.org/jdk/pull/22386/files/3bae7423063688c1fe40ab3c1d5b50c2b94a8229)
 * [Fei Yang](https://openjdk.org/census#fyang) (@RealFYang - **Reviewer**)
 * [Ludovic Henry](https://openjdk.org/census#luhenry) (@luhenry - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/22386/head:pull/22386` \
`$ git checkout pull/22386`

Update a local copy of the PR: \
`$ git checkout pull/22386` \
`$ git pull https://git.openjdk.org/jdk.git pull/22386/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 22386`

View PR using the GUI difftool: \
`$ git pr show -t 22386`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/22386.diff">https://git.openjdk.org/jdk/pull/22386.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/22386#issuecomment-2500061891)
</details>

### CODE REVIEW

File: src/hotspot/cpu/riscv/macroAssembler_riscv.cpp [hunk 1]
```diff
@@ -1128,6 +1128,147 @@ void MacroAssembler::wrap_label(Register r1, Register r2, Label &L,
...
+    xorr(t0, cmp1, cmp2);
+    czero_eqz(dst, dst, t0);
+    czero_nez(t0 , src, t0);
+    orr(dst, dst, t0);
+    return;
+  }
+  Label no_set;
+  bne(cmp1, cmp2, no_set);
+  mv(dst, src);
+  bind(no_set);
```
Comment (Hamlin-Li) [inline src/hotspot/cpu/riscv/macroAssembler_riscv.cpp:1143]: Do we have performance data comparison between `-UseZicond ` and `+ UseZicond `?

File: src/hotspot/cpu/riscv/c2_MacroAssembler_riscv.cpp [hunk 2]
```diff
@@ -2003,10 +2003,48 @@ void C2_MacroAssembler::enc_cmpEqNe_imm0_branch(int cmpFlag, Register op1, Label
...
+      if (is_unsigned) {
+        cmov_gtu(op1, op2, dst, src);
+      } else {
+        cmov_gt(op1, op2, dst, src);
+      }
+      break;
+    default:
+      assert(false, "unsupported compare condition");
+      ShouldNotReachHere();
+  }
```
Comment (Hamlin-Li) [inline src/hotspot/cpu/riscv/c2_MacroAssembler_riscv.cpp:2047]: Seems we can use the similar way in `C2_MacroAssembler::cmp_branch` +`conditional_branches` to simplify the code.

**Review by Hamlin-Li — COMMENTED** [review]
Thanks for adding the support for this extension!
The code change looks good. Just have some questions below.

File: src/hotspot/cpu/riscv/macroAssembler_riscv.cpp (hunk 1 above)
Comment (Hamlin-Li) [reply src/hotspot/cpu/riscv/macroAssembler_riscv.cpp:1143]: Seems the critical path when `+UseZicond` is longer than `-UseZicond`, I'm not sure how much impact the branch code will bring when `-UseZicond`.

File: src/hotspot/cpu/riscv/c2_MacroAssembler_riscv.cpp (hunk 2 above)
Comment (robehn) [reply src/hotspot/cpu/riscv/c2_MacroAssembler_riscv.cpp:2047]: The "BoolTest" is a C2 construct, C1 uses LIR_Condition, there is no generic test enum.
So you would need one table to translate from C2 -> masm and one from C1 -> masm, and then one table to pick the correct version.

You can turn any switch statement into a table with function pointers, but:
`cmp_branch(cmpFlag ^ (1 << neg_cond_bits), op1, op2, L);`
Trying to figure out which compare is selected is shoter, yes, but much more complex IMHO.

Comment (Hamlin-Li) [reply src/hotspot/cpu/riscv/c2_MacroAssembler_riscv.cpp:2047]: I see, thanks for clarification.

**Review by Hamlin-Li — APPROVED** [review]
Looks good to me. Thanks!

**Review by RealFYang — COMMENTED** [review]
Hi, thanks for this work! I once tried to find this extension but I think I missed it :-)
Seems that we need two special AD match rules so that `eq`/`ne` zero cases could be further optimized.
```
Conditional select, if zero
 rd = (rc == 0) ? rs1 : rs2
=>
 czero.nez  rd, rs1, rc
 czero.eqz  rtmp, rs2, rc
 or         rd, rd, rtmp
```

```
Conditional select, if non-zero
 rd = (rc != 0) ? rs1 : rs2
=>
czero.eqz  rd, rs1, rc
czero.nez  rtmp, rs2, rc
or         rd, rd, rtmp
```

File: src/hotspot/cpu/riscv/c2_MacroAssembler_riscv.cpp [hunk 3]
```diff
@@ -2003,10 +2003,48 @@ void C2_MacroAssembler::enc_cmpEqNe_imm0_branch(int cmpFlag, Register op1, Label
 }
 
 void C2_MacroAssembler::enc_cmove(int cmpFlag, Register op1, Register op2, Register dst, Register src) {
-  Label L;
-  cmp_branch(cmpFlag ^ (1 << neg_cond_bits), op1, op2, L);
```
[the remark this answers is not in the document]
Comment (robehn) [reply src/hotspot/cpu/riscv/c2_MacroAssembler_riscv.cpp:2007]: Thanks

**Comment by robehn** [comment]
> Hi, thanks for this work! I once tried to find this extension but I think I missed it :-) Seems that we need two special AD match rules to catch `eq`/`ne` zero cases could be further optimized. (Say, we use `t0` for `rtmp` in this code sequence)
> 
> ```
> Conditional select, if zero
>  rd = (rc == 0) ? rs1 : rs2
> =>
>  czero.nez  rd, rs1, rc
>  czero.eqz  rtmp, rs2, rc
>  or         rd, rd, rtmp
> ```
> 
> ```
> Conditional select, if non-zero
>  rd = (rc != 0) ? rs1 : rs2
> =>
> czero.eqz  rd, rs1, rc
> czero.nez  rtmp, rs2, rc
> or         rd, rd, rtmp
> ```

Thanks! Yes we should look into that.

**Review by RealFYang — APPROVED** [review]
Update change looks good. I am OK if we want to leave other optimization opportunities to some follow-up PR. Up to you!

**Review by luhenry — APPROVED** [review]

**Comment by luhenry** [comment]
I love the added the new `test/hotspot/gtest/riscv/test_assembler_riscv.cpp`. Is it going to be picked up, compiled and run automatically as well?

**Comment by robehn** [comment]
> I love the added the new `test/hotspot/gtest/riscv/test_assembler_riscv.cpp`. Is it going to be picked up, compiled and run automatically as well?

Yes:
```
[rehn@rehn-black open]$ make test-gtest CONF=linux-riscv64-server-fastdebug | grep Risc
[----------] 1 test from RiscV
[ RUN      ] RiscV.cmov_vm
[       OK ] RiscV.cmov_vm (2 ms)
[----------] 1 test from RiscV (2 ms total)

```

**Comment by robehn** [comment]
> Update change looks good. I am OK if we want to leave other optimization opportunities to some follow-up PR. Up to you!

Let's do that, thanks!

File: src/hotspot/cpu/riscv/macroAssembler_riscv.cpp (hunk 1 above)
Comment (robehn) [reply src/hotspot/cpu/riscv/macroAssembler_riscv.cpp:1143]: Yes, exactly, a branch is always more 'intrusive' in the pipeline than plain scalar ops.
As we discussed, there are cases where czero perform in pair with branch, but in general it's a win.

**Pull request closed without merging**